File: frontend/src/data/hackathones.ts
// Superficie de HACKATONES contra la API real (no mocks). Como `finance.ts`: funciones async + un
// transform snake_case → camelCase. El dashboard trae las filas crudas y ordena/filtra en el
// cliente.

import { apiGet } from "@/lib/api"
import type { Hackathon, HackathonModality } from "@/types/domain"

interface HackathonApiRow {
  id: number
  name: string
  starts_on: string | null
  ends_on: string | null
  registration_deadline: string | null
  modality: string
  location: string
  url: string
  organizer: string
  technologies: string
  prizes: string
  requirements: string
  description: string
  evidence: string
  source_inbox_ids: number[]
  created_at: string
}

interface HackathonApiList {
  items: HackathonApiRow[]
  next_cursor: number | null
}

function toHackathon(r: HackathonApiRow): Hackathon {
  return {
    id: r.id,
    name: r.name,
    startsOn: r.starts_on,
    endsOn: r.ends_on,
    registrationDeadline: r.registration_deadline,
    modality: r.modality as HackathonModality,
    location: r.location,
    url: r.url,
    organizer: r.organizer,
    technologies: r.technologies,
    prizes: r.prizes,
    requirements: r.requirements,
    description: r.description,
    evidence: r.evidence,
    sourceInboxIds: r.source_inbox_ids,
    createdAt: r.created_at,
  }
}

export interface FetchHackathonesOpts {
  modality?: string
  /** starts_on >= since (YYYY-MM-DD) */
  since?: string
  /** starts_on < until (YYYY-MM-DD) */
  until?: string
  /** Tope total de filas a traer (paginando). */
  max?: number
}

/**
 * Todos los hackatones del usuario (GET /hackathones/events), paginando por cursor igual que
 * `fetchFinanceTransactions`. El dashboard ordena/filtra en el cliente, así que por defecto trae todo.
 */
export async function fetchHackathones(opts?: FetchHackathonesOpts): Promise<Hackathon[]> {
  const max = opts?.max ?? 5000
  const pageSize = 500
  const out: Hackathon[] = []
  let cursor: number | null = null
  while (out.length < max) {
    const qs = new URLSearchParams()
    if (opts?.modality) qs.set("modality", opts.modality)
    if (opts?.since) qs.set("since", opts.since)
    if (opts?.until) qs.set("until", opts.until)
    qs.set("limit", String(pageSize))
    if (cursor != null) qs.set("cursor", String(cursor))
    const page = await apiGet<HackathonApiList>(`/hackathones/events?${qs.toString()}`)
    out.push(...page.items.map(toHackathon))
    if (page.next_cursor == null || page.items.length === 0) break
    cursor = page.next_cursor
  }
  return out
}

File: src/memex/api/routers/hackathones.py
from datetime import date
from typing import Annotated, Any

from fastapi import APIRouter, Depends, Query
from sqlalchemy import text

from memex.api.auth import current_user_id
from memex.api.schemas import HackathonList
from memex.db import connection
from memex.logging import get_logger

router = APIRouter(prefix="/hackathones", tags=["hackathones"])

UserID = Annotated[int, Depends(current_user_id)]

_log = get_logger("memex.api.hackathones")


@router.get("/events", response_model=HackathonList)
async def list_hackathones(
    user_id: UserID,
    modality: str | None = None,
    since: date | None = None,
    until: date | None = None,
    limit: Annotated[int, Query(ge=1, le=500)] = 200,
    cursor: int | None = Query(default=None, description="id > cursor for pagination"),
) -> dict[str, Any]:
    """Lista los hackatones extraídos por el módulo hackathones (tabla `mod_hackathones_events`).

    Devuelve filas crudas; el dashboard agrega/ordena en el cliente. La paginación es por cursor
    (`id > :cur`) igual que `/inbox` y finance. Los filtros `modality`/`since`/`until`
    (sobre `starts_on`) espejan el patrón de finance; las filas sin fecha quedan fuera al filtrar
    por rango (es el comportamiento esperado: filtrar por fecha pide tener fecha).
    """
    where: list[str] = ["user_id = :uid"]
    params: dict[str, Any] = {"uid": user_id, "limit": limit}

    if modality is not None:
        where.append("modality = :modality")
        params["modality"] = modality
    if since is not None:
        where.append("starts_on >= :since")
        params["since"] = since
    if until is not None:
        where.append("starts_on < :until")
        params["until"] = until
    if cursor is not None:
        where.append("id > :cur")
        params["cur"] = cursor

    sql = f"""
        SELECT id, name, starts_on, ends_on, registration_deadline, modality, location, url,
               organizer, technologies, prizes, requirements, description, evidence,
               source_inbox_ids, created_at
        FROM mod_hackathones_events
        WHERE {" AND ".join(where)}
        ORDER BY id
        LIMIT :limit
    """
    with connection() as conn:
        rows = conn.execute(text(sql), params).mappings().all()

    items = [dict(r) for r in rows]
    next_cursor = items[-1]["id"] if len(items) == limit else None
    _log.info("hackathones.events.listed", user_id=user_id, count=len(items), modality=modality)
    return {"items": items, "next_cursor": next_cursor}
